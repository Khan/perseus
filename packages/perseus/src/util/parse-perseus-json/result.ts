import invariant from "tiny-invariant";

export type Result<S, F> = Success<S> | Failure<F>;

export type Success<T> = {type: "success"; value: T};
export type Failure<T> = {type: "failure"; detail: T};

export function success<T>(value: T): Success<T> {
    return {type: "success", value};
}

export function failure<T>(detail: T): Failure<T> {
    return {type: "failure", detail};
}

export function isFailure<S, F>(result: Result<S, F>): result is Failure<F> {
    return result.type === "failure";
}

export function isSuccess<S, F>(result: Result<S, F>): result is Success<S> {
    return result.type === "success";
}

export function assertFailure<S, F>(
    result: Result<S, F>,
): asserts result is Failure<F> {
    invariant(
        isFailure(result),
        "expected result to be a Failure, but got a Success",
    );
}

export function assertSuccess<S, F>(
    result: Result<S, F>,
): asserts result is Success<S> {
    invariant(
        isSuccess(result),
        "expected result to be a Success, but got a Failure",
    );
}

// Result's `all` function is similar to Promise.all: given an array of
// results, it returns success if all succeeded, and failure if any failed.
export function all<S, F>(
    results: Array<Result<S, F>>,
    combineFailureDetails: (a: F, b: F) => F = (a) => a,
): Result<S[], F> {
    const values: S[] = [];
    const failureDetails: F[] = [];
    for (const result of results) {
        if (result.type === "success") {
            values.push(result.value);
        } else {
            failureDetails.push(result.detail);
        }
    }

    if (failureDetails.length > 0) {
        return failure(failureDetails.reduce(combineFailureDetails));
    }
    return success(values);
}
