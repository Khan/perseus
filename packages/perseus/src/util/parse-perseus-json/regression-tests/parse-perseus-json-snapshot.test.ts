import * as fs from "fs";
import {join} from "path";

import {parseAndTypecheckPerseusItem} from "../index";
import {assertSuccess} from "../result";

const dataFiles = fs.readdirSync(join(__dirname, "data"));

// If you change the parsers to migrate/convert data to a new format, you may
// have to regenerate these snapshots. That's expected and okay!

describe("parseAndTypecheckPerseusItem", () => {
    it.each(dataFiles)("correctly parses data/%s", (filename) => {
        const json = fs.readFileSync(
            join(__dirname, "data", filename),
            "utf-8",
        );
        const result = parseAndTypecheckPerseusItem(json);
        assertSuccess(result);
        expect(result.value).toMatchSnapshot();
    });
});
