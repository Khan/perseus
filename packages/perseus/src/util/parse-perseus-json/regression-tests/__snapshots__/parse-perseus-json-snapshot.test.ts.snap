// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseAndTypecheckPerseusItem correctly parses data/radio-missing-noneOfTheAbove.json 1`] = `
{
  "answer": undefined,
  "answerArea": {
    "calculator": false,
    "chi2Table": false,
    "periodicTable": false,
    "tTable": false,
    "zTable": false,
  },
  "hints": [],
  "itemDataVersion": {
    "major": 0,
    "minor": 1,
  },
  "question": {
    "content": "$\\begin{align} 
A &= \\{x: x \\text{ is a rectangle in the plane P} \\} \\\\\\\\
B &= \\{x: x \\text{ is a square in the plane P} \\}
\\end{align}$

**Is $A \\subset B$ ?**

[[â˜ƒ radio 1]]
",
    "images": {},
    "metadata": undefined,
    "widgets": {
      "radio 1": {
        "alignment": "default",
        "graded": true,
        "key": undefined,
        "options": {
          "choices": [
            {
              "clue": undefined,
              "content": "Yes",
              "correct": false,
              "isNoneOfTheAbove": undefined,
              "widgets": undefined,
            },
            {
              "clue": undefined,
              "content": "No",
              "correct": true,
              "isNoneOfTheAbove": undefined,
              "widgets": undefined,
            },
          ],
          "countChoices": false,
          "deselectEnabled": false,
          "displayCount": null,
          "hasNoneOfTheAbove": false,
          "multipleSelect": false,
          "noneOfTheAbove": undefined,
          "onePerLine": undefined,
          "randomize": false,
        },
        "static": false,
        "type": "radio",
        "version": {
          "major": 1,
          "minor": 0,
        },
      },
    },
  },
}
`;
